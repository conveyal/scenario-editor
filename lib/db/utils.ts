import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollectionResults(obj: unknown) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollectionResults)
  if (typeof obj === 'object') return mapValues(obj, serializeCollectionResults)
  return obj
}
