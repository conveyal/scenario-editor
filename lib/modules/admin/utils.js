import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect(Component) {
  return connect(createStructuredSelector(select), actions)(Component)
}

const add0 = n => (n >= 10 ? n : `0${n}`)

export function msToDuration(ms) {
  const durationS = Math.floor((Date.now() - ms) / 1000)
  const s = durationS % 60
  const m = Math.floor((durationS / 60) % 60)
  const h = Math.floor((durationS / 60 / 60) % 60)

  return `${add0(h)}:${add0(m)}:${add0(s)}`
}

/**
 * Example commit string: "v3.4.0-118-gbe7199c". Take the last section after
 * splitting on the hyphens and remove the "g".
 */
const githubUrl = 'https://github.com/conveyal/r5/commits'
export function createR5Url(workerCommit) {
  // not a release version
  if (workerCommit.indexOf('-') !== -1) {
    const hash = workerCommit.split('-')[2].slice(1) //
    return `${githubUrl}/${hash}`
  }
  return `${githubUrl}/${workerCommit}`
}

export function createWorkerUrl(instanceId, region) {
  return `https://${region}.console.aws.amazon.com/ec2/v2/home?region=${region}#Instances:instanceId=${instanceId};sort=tag:Name`
}

const C1 = '#2389c9'
const C2 = '#36c923'
const C3 = '#aab4bb'

/**
 * Create a progress bar with a linear-gradient.
 */
export function createLinearGradientForJob(c, d, t) {
  const p1 = (c / t) * 100
  const p2 = (d / t) * 100
  return `linear-gradient(to right,
    ${C1} 0%,
    ${C1} ${p1}%,
    ${C2} ${p1}%,
    ${C2} ${p2}%,
    ${C3} ${p2}%,
    ${C3} 100%)`
}
