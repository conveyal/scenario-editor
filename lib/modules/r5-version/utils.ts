import {
  MINIMUM_R5_VERSION,
  VERSION_PARSE_REGEX,
  OLD_VERSION_PARSE_REGEX
} from './constants'

/**
 * Convert an r5 version to a number for comparing
 * NB Assumes no major, minor, patch, or commit version exceeds 100
 */
export function versionToNumber(version?: string): number {
  const versionString = version || ''
  if (versionString.match(/\./g) && versionString.match(/\./g).length == 1) {
    const versionMatch = versionString.match(VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [, major = '0', minor = '0', commit = '0'] = versionMatch

      return (
        parseInt(major) * 10000 + parseInt(minor) * 100 + parseInt(commit) / 100
      )
    }

    return 0
  } else {
    const versionMatch = versionString.match(OLD_VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [
        ,
        major = '0',
        minor = '0',
        patch = '0',
        commit = '0'
      ] = versionMatch

      return (
        parseInt(major) * 10000 +
        parseInt(minor) * 100 +
        parseInt(patch) +
        parseInt(commit) / 100
      )
    }

    return 0
  }
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = (version || '').search(/-/g) !== -1
  return isTestVersion || workerVersionInRange(version, min, max)
}
