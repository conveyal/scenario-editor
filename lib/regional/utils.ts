function getDefaultCutoff(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.cutoffsMinutes))
    return a.cutoffsMinutes[Math.floor(a.cutoffsMinutes.length / 2)]
  return a.cutoffMinutes
}

function getDefaultDestinationPointSet(a: CL.RegionalAnalysis): string | void {
  const ids = a.destinationPointSetIds
  if (Array.isArray(ids)) return ids[0]
}

function getDefaultPercentile(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.travelTimePercentiles))
    return a.travelTimePercentiles[
      Math.floor(a.travelTimePercentiles.length / 2)
    ]
  return a.travelTimePercentile
}

/**
 * Choose
 */
export function getDefaultVariants(analysis: CL.RegionalAnalysis) {
  return {
    cutoff: getDefaultCutoff(analysis),
    destinationPointSetId: getDefaultDestinationPointSet(analysis),
    percentile: getDefaultPercentile(analysis)
  }
}

export function getComparisonVariants(comparison: CL.RegionalAnalysis) {
  return {
    comparisonCutoff: getDefaultCutoff(comparison),
    comparisonDestinationPointSetId: getDefaultDestinationPointSet(comparison),
    comparisonPercentile: getDefaultPercentile(comparison)
  }
}

/**
 * Don't compare incomparable analyses.
 */
export function getValidComparisonAnalyses(
  allAnalyses: CL.RegionalAnalysis[],
  activeJobs: CL.RegionalJob[],
  analysis: CL.RegionalAnalysis
) {
  return allAnalyses.filter(
    (c) =>
      activeJobs.findIndex((job) => job.jobId === c._id) === -1 &&
      c.request.originPointSetKey == null &&
      c.zoom === analysis.zoom &&
      c.width === analysis.width &&
      c.height === analysis.height &&
      c.north === analysis.north &&
      c.west === analysis.west
  )
}
