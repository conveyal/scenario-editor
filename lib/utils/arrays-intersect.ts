// Do the arrays intersect?
export default function arraysIntersect<T>(a1: T[], a2: T[]) {
  for (const v of a1) {
    if (a2.indexOf(v) > -1) return true
  }
  return false
}
