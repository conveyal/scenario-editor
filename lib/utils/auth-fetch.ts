import {parse} from 'cookie'

import {safeFetch, SafeResponse} from './safe-fetch'

export function createAuthHeaders(user?: CL.User) {
  const headers = {
    Authorization: `bearer ${user?.idToken}`
  }

  // Add the admin access group for administrators if it exists.
  if (user?.accessGroup === process.env.NEXT_PUBLIC_ADMIN_ACCESS_GROUP) {
    const adminTempAccessGroup = parse(document.cookie).adminTempAccessGroup
    if (adminTempAccessGroup?.length > 0) {
      headers['X-Conveyal-Access-Group'] = adminTempAccessGroup
    }
  }
  return headers
}

/**
 * Fetch wrapper that includes authentication. Defaults headers to JSON.
 */
export default function authFetch<T>(
  url: string,
  user: CL.User,
  parser: (res: Response) => Promise<T> = (res) => res.json(),
  options?: RequestInit
): Promise<SafeResponse<T>> {
  return safeFetch(url, parser, {
    mode: 'cors',
    ...options,
    headers: {
      ...createAuthHeaders(user),
      ...options?.headers
    }
  })
}
