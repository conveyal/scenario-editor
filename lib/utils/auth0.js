// @flow
import message from '@conveyal/woonerf/message'
import Auth0Lock from 'auth0-lock'

const CLIENT_ID = process.env.AUTH0_CLIENT_ID
const DOMAIN = process.env.AUTH0_DOMAIN

export const authIsRequired = CLIENT_ID && DOMAIN

const localStorage = window && window.localStorage
  ? window.localStorage
  : {
    getItem () {},
    setItem () {}
  }

const lock = authIsRequired
  ? new Auth0Lock(CLIENT_ID, DOMAIN, {
    auth: {
      params: {
        scope: 'openid email profile analyst offline_access'
      },
      redirect: false
    },
    allowSignUp: false,
    theme: {
      logo: 'https://s3-eu-west-1.amazonaws.com/analyst-logos/conveyal-128x128.png',
      primaryColor: '#2389c9'
    },
    closable: false,
    languageDictionary: {
      title: message('authentication.logIn')
    }
  })
  : {
    checkSession () {},
    getUserInfo () {},
    hide () {},
    logout () {},
    on () {},
    show () {}
  }

function flashError (error) {
  console.error(error)
  lock.show({
    flashMessage: {
      type: 'error',
      text: error.error_description
    }
  })
}

// Set the default handler for auth errors
lock.on('authorization_error', flashError)
lock.on('unrecoverable_error', flashError)

export function login (next: (any) => void) {
  function handleAuthResult (authResult) {
    console.log('handleAuthResult', authResult)
    lock.getUserInfo(authResult.accessToken, (error, profile) => {
      if (error) {
        localStorage.removeItem('user')
        lock.show({
          flashMessage: {
            type: 'error',
            text: error.error_description
          }
        })
      } else {
        localStorage.setItem('user', JSON.stringify(profile))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      }
    })
  }

  lock.checkSession({}, (error, authResult) => {
    console.log('checkSession', error, authResult)
    if (error || !authResult) {
      lock.show()
      lock.on('authenticated', handleAuthResult)
    } else {
      try {
        const profile = JSON.parse(localStorage.getItem('user'))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      } catch (e) {
        console.error('error processing user profile from localStorage', e)
        handleAuthResult(authResult)
      }
    }
  })
}

export function logout () {
  localStorage.removeItem('user')
  lock.logout({returnTo: `${window.location.origin}/login`})
}
