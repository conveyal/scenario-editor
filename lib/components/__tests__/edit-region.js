// @flow
import React from 'react'

import EditRegion from '../edit-region'
import {mockRegion, mockWithProvider} from '../../utils/mock-data'

describe('Component > EditRegion', () => {
  it('renders correctly', () => {
    const props = {
      region: mockRegion,
      clearUncreatedRegion: jest.fn(),
      fetch: jest.fn(),
      save: jest.fn(),
      load: jest.fn(),
      create: jest.fn(),
      deleteRegion: jest.fn(),
      setLocally: jest.fn(),
      setCenter: jest.fn(),
      isEditing: true
    }

    // mount component
    const {snapshot} = mockWithProvider(<EditRegion {...props} />)
    expect(snapshot()).toMatchSnapshot()
    const noCalls = [
      'clearUncreatedRegion',
      'save',
      'load',
      'setLocally',
      'setCenter'
    ]
    noCalls.forEach(fn => {
      expect(props[fn]).not.toBeCalled()
    })
  })
})
