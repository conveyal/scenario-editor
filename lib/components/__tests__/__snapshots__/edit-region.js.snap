// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component > EditRegion renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <EditRegion
    clearUncreatedRegion={[MockFunction]}
    create={[MockFunction]}
    deleteRegion={[MockFunction]}
    fetch={[MockFunction]}
    isEditing={true}
    load={[MockFunction]}
    region={
      Object {
        "_id": "1",
        "accessGroup": "accessGroupName",
        "bounds": Object {
          "east": -76,
          "north": 39,
          "south": 38,
          "west": -77,
        },
        "bundles": Array [
          Object {
            "_id": "1",
            "accessGroup": "accessGroupName",
            "centerLat": 38.90124,
            "centerLon": -77.015615,
            "createdAt": "2017-11-08T12:30:50.133Z",
            "createdBy": "person@gmail.com",
            "errorCode": "",
            "feeds": Array [],
            "name": "Mock Bundle",
            "nonce": "12345",
            "regionId": "1",
            "status": "DONE",
            "totalFeeds": 0,
            "updatedAt": "2017-11-08T12:31:19.602Z",
            "updatedBy": "person@gmail.com",
          },
        ],
        "createdAt": "2017-11-08T12:30:50.133Z",
        "createdBy": "person@gmail.com",
        "description": "Region description",
        "name": "Mock Region",
        "nonce": "12345",
        "opportunityDatasets": Array [
          Object {
            "_id": "Workers_total",
            "name": "Workers total",
            "sourceName": "lodes-data-2014",
          },
        ],
        "statusCode": "DONE",
        "updatedAt": "2017-11-08T12:31:19.602Z",
        "updatedBy": "person@gmail.com",
      }
    }
    save={[MockFunction]}
    setCenter={[MockFunction]}
    setLocally={[MockFunction]}
  >
    <Application
      map={[Function]}
    >
      <div>
        <Connect(ErrorModal)>
          <ErrorModal
            clearError={[Function]}
          >
            <span />
          </ErrorModal>
        </Connect(ErrorModal)>
        <Connect(Sidebar)>
          <Sidebar
            currentPath="/"
            dispatch={[Function]}
            outstandingRequests={0}
            projectId="1"
            regionId="1"
          >
            <div
              className="Sidebar"
            >
              <div
                className="Sidebar-logo"
              />
              <SidebarNavItem
                href="/"
                icon="globe"
                text="Regions"
              >
                <Tip
                  className="Sidebar-navItem"
                  tip="Regions"
                >
                  <span
                    className="Tip Sidebar-navItem"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                  >
                    <Link
                      className="Sidebar-navItem-contents"
                      onlyActiveOnIndex={false}
                      style={Object {}}
                      to="/"
                    >
                      <a
                        className="Sidebar-navItem-contents"
                        onClick={[Function]}
                        style={Object {}}
                      >
                        <ItemContents
                          icon="globe"
                          text="Regions"
                        >
                          <span>
                            <Icon
                              type="globe"
                            >
                              <i
                                className="fa fa-globe fa-fw "
                              />
                            </Icon>
                            <span
                              className="Sidebar-navItem-text"
                            >
                               
                              Regions
                            </span>
                          </span>
                        </ItemContents>
                      </a>
                    </Link>
                  </span>
                </Tip>
              </SidebarNavItem>
              <div>
                <SidebarNavItem
                  active={null}
                  href="/regions/1/edit"
                  icon="map-o"
                  text="Region Settings"
                >
                  <Tip
                    className="Sidebar-navItem"
                    tip="Region Settings"
                  >
                    <span
                      className="Tip Sidebar-navItem"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <Link
                        className="Sidebar-navItem-contents"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/regions/1/edit"
                      >
                        <a
                          className="Sidebar-navItem-contents"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <ItemContents
                            icon="map-o"
                            text="Region Settings"
                          >
                            <span>
                              <Icon
                                type="map-o"
                              >
                                <i
                                  className="fa fa-map-o fa-fw "
                                />
                              </Icon>
                              <span
                                className="Sidebar-navItem-text"
                              >
                                 
                                Region Settings
                              </span>
                            </span>
                          </ItemContents>
                        </a>
                      </Link>
                    </span>
                  </Tip>
                </SidebarNavItem>
                <SidebarNavItem
                  active={
                    Array [
                      "/",
                    ]
                  }
                  href="/regions/1"
                  icon="cubes"
                  text="Projects"
                >
                  <Tip
                    className="Sidebar-navItem active"
                    tip="Projects"
                  >
                    <span
                      className="Tip Sidebar-navItem active"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <Link
                        className="Sidebar-navItem-contents"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/regions/1"
                      >
                        <a
                          className="Sidebar-navItem-contents"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <ItemContents
                            icon="cubes"
                            text="Projects"
                          >
                            <span>
                              <Icon
                                type="cubes"
                              >
                                <i
                                  className="fa fa-cubes fa-fw "
                                />
                              </Icon>
                              <span
                                className="Sidebar-navItem-text"
                              >
                                 
                                Projects
                              </span>
                            </span>
                          </ItemContents>
                        </a>
                      </Link>
                    </span>
                  </Tip>
                </SidebarNavItem>
                <SidebarNavItem
                  active={null}
                  href="/regions/1/bundles"
                  icon="database"
                  text="GTFS Bundles"
                >
                  <Tip
                    className="Sidebar-navItem"
                    tip="GTFS Bundles"
                  >
                    <span
                      className="Tip Sidebar-navItem"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <Link
                        className="Sidebar-navItem-contents"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/regions/1/bundles"
                      >
                        <a
                          className="Sidebar-navItem-contents"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <ItemContents
                            icon="database"
                            text="GTFS Bundles"
                          >
                            <span>
                              <Icon
                                type="database"
                              >
                                <i
                                  className="fa fa-database fa-fw "
                                />
                              </Icon>
                              <span
                                className="Sidebar-navItem-text"
                              >
                                 
                                GTFS Bundles
                              </span>
                            </span>
                          </ItemContents>
                        </a>
                      </Link>
                    </span>
                  </Tip>
                </SidebarNavItem>
                <SidebarNavItem
                  active={null}
                  href="/regions/1/opportunities"
                  icon="th"
                  text="Opportunity Datasets"
                >
                  <Tip
                    className="Sidebar-navItem"
                    tip="Opportunity Datasets"
                  >
                    <span
                      className="Tip Sidebar-navItem"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <Link
                        className="Sidebar-navItem-contents"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/regions/1/opportunities"
                      >
                        <a
                          className="Sidebar-navItem-contents"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <ItemContents
                            icon="th"
                            text="Opportunity Datasets"
                          >
                            <span>
                              <Icon
                                type="th"
                              >
                                <i
                                  className="fa fa-th fa-fw "
                                />
                              </Icon>
                              <span
                                className="Sidebar-navItem-text"
                              >
                                 
                                Opportunity Datasets
                              </span>
                            </span>
                          </ItemContents>
                        </a>
                      </Link>
                    </span>
                  </Tip>
                </SidebarNavItem>
                <SidebarNavItem
                  active={null}
                  href="/regions/1/projects/1"
                  icon="pencil"
                  text="Edit Modifications"
                >
                  <Tip
                    className="Sidebar-navItem"
                    tip="Edit Modifications"
                  >
                    <span
                      className="Tip Sidebar-navItem"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <Link
                        className="Sidebar-navItem-contents"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/regions/1/projects/1"
                      >
                        <a
                          className="Sidebar-navItem-contents"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <ItemContents
                            icon="pencil"
                            text="Edit Modifications"
                          >
                            <span>
                              <Icon
                                type="pencil"
                              >
                                <i
                                  className="fa fa-pencil fa-fw "
                                />
                              </Icon>
                              <span
                                className="Sidebar-navItem-text"
                              >
                                 
                                Edit Modifications
                              </span>
                            </span>
                          </ItemContents>
                        </a>
                      </Link>
                    </span>
                  </Tip>
                </SidebarNavItem>
                <SidebarNavItem
                  active={null}
                  href="/regions/1/analysis"
                  icon="area-chart"
                  text="Analyze"
                >
                  <Tip
                    className="Sidebar-navItem"
                    tip="Analyze"
                  >
                    <span
                      className="Tip Sidebar-navItem"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <Link
                        className="Sidebar-navItem-contents"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/regions/1/analysis"
                      >
                        <a
                          className="Sidebar-navItem-contents"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <ItemContents
                            icon="area-chart"
                            text="Analyze"
                          >
                            <span>
                              <Icon
                                type="area-chart"
                              >
                                <i
                                  className="fa fa-area-chart fa-fw "
                                />
                              </Icon>
                              <span
                                className="Sidebar-navItem-text"
                              >
                                 
                                Analyze
                              </span>
                            </span>
                          </ItemContents>
                        </a>
                      </Link>
                    </span>
                  </Tip>
                </SidebarNavItem>
                <SidebarNavItem
                  active={null}
                  href="/regions/1/regional"
                  icon="server"
                  text="Regional Analysis"
                >
                  <Tip
                    className="Sidebar-navItem"
                    tip="Regional Analysis"
                  >
                    <span
                      className="Tip Sidebar-navItem"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <Link
                        className="Sidebar-navItem-contents"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/regions/1/regional"
                      >
                        <a
                          className="Sidebar-navItem-contents"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <ItemContents
                            icon="server"
                            text="Regional Analysis"
                          >
                            <span>
                              <Icon
                                type="server"
                              >
                                <i
                                  className="fa fa-server fa-fw "
                                />
                              </Icon>
                              <span
                                className="Sidebar-navItem-text"
                              >
                                 
                                Regional Analysis
                              </span>
                            </span>
                          </ItemContents>
                        </a>
                      </Link>
                    </span>
                  </Tip>
                </SidebarNavItem>
              </div>
              <div
                className="Sidebar-bottom"
              >
                <SidebarNavItem
                  href="http://docs.analysis.conveyal.com/"
                  icon="question-circle"
                  text="Help and Documentation"
                >
                  <Tip
                    className="Sidebar-navItem"
                    tip="Help and Documentation"
                  >
                    <span
                      className="Tip Sidebar-navItem"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <a
                        className="Sidebar-navItem-contents"
                        href="http://docs.analysis.conveyal.com/"
                        target="_blank"
                      >
                        <ItemContents
                          icon="question-circle"
                          text="Help and Documentation"
                        >
                          <span>
                            <Icon
                              type="question-circle"
                            >
                              <i
                                className="fa fa-question-circle fa-fw "
                              />
                            </Icon>
                            <span
                              className="Sidebar-navItem-text"
                            >
                               
                              Help and Documentation
                            </span>
                          </span>
                        </ItemContents>
                      </a>
                    </span>
                  </Tip>
                </SidebarNavItem>
              </div>
            </div>
          </Sidebar>
        </Connect(Sidebar)>
        <div
          className="Fullscreen"
        >
          <Connect(Map)>
            <Map
              bounds={
                Object {
                  "_northEast": Object {
                    "lat": 37.365,
                    "lng": -121.924,
                  },
                  "_southWest": Object {
                    "lat": 36.9707,
                    "lng": -122.0279,
                  },
                }
              }
              center={
                Object {
                  "lat": 38.8886,
                  "lon": -77.043,
                }
              }
              dispatch={[Function]}
              maxBounds={
                Object {
                  "_northEast": Object {
                    "lat": 90,
                    "lng": 200,
                  },
                  "_southWest": Object {
                    "lat": -90,
                    "lng": -200,
                  },
                }
              }
              minZoom={2}
              tileUrl="mock.url/tile"
              zIndex={-10}
              zoom={12}
            >
              <Map
                animate={false}
                bounds={
                  Object {
                    "_northEast": Object {
                      "lat": 37.365,
                      "lng": -121.924,
                    },
                    "_southWest": Object {
                      "lat": 36.9707,
                      "lng": -122.0279,
                    },
                  }
                }
                center={
                  Object {
                    "lat": 38.8886,
                    "lon": -77.043,
                  }
                }
                maxBounds={
                  Object {
                    "_northEast": Object {
                      "lat": 90,
                      "lng": 200,
                    },
                    "_southWest": Object {
                      "lat": -90,
                      "lng": -200,
                    },
                  }
                }
                minZoom={2}
                useFlyTo={false}
                zoom={12}
              >
                <div>
                  <TileLayer
                    attribution="<a href=\\"http://www.openstreetmap.org/copyright\\">OpenStreetMap</a> contributors, &copy; <a href='https://www.mapbox.com/about/maps/'>Mapbox</a>"
                    url="mock.url/tile"
                    zIndex={-10}
                  />
                  <EditBounds
                    bounds={
                      Object {
                        "east": -76,
                        "north": 39,
                        "south": 38,
                        "west": -77,
                      }
                    }
                    save={[Function]}
                  >
                    <g>
                      <Marker
                        draggable={true}
                        onDragEnd={[Function]}
                        position={
                          Array [
                            38,
                            -77,
                          ]
                        }
                      >
                        <g
                          data-props={
                            Object {
                              "draggable": true,
                              "onDragEnd": [Function],
                              "position": Array [
                                38,
                                -77,
                              ],
                            }
                          }
                          type="Marker"
                        />
                      </Marker>
                      <Marker
                        draggable={true}
                        onDragEnd={[Function]}
                        position={
                          Array [
                            39,
                            -77,
                          ]
                        }
                      >
                        <g
                          data-props={
                            Object {
                              "draggable": true,
                              "onDragEnd": [Function],
                              "position": Array [
                                39,
                                -77,
                              ],
                            }
                          }
                          type="Marker"
                        />
                      </Marker>
                      <Marker
                        draggable={true}
                        onDragEnd={[Function]}
                        position={
                          Array [
                            39,
                            -76,
                          ]
                        }
                      >
                        <g
                          data-props={
                            Object {
                              "draggable": true,
                              "onDragEnd": [Function],
                              "position": Array [
                                39,
                                -76,
                              ],
                            }
                          }
                          type="Marker"
                        />
                      </Marker>
                      <Marker
                        draggable={true}
                        onDragEnd={[Function]}
                        position={
                          Array [
                            38,
                            -76,
                          ]
                        }
                      >
                        <g
                          data-props={
                            Object {
                              "draggable": true,
                              "onDragEnd": [Function],
                              "position": Array [
                                38,
                                -76,
                              ],
                            }
                          }
                          type="Marker"
                        />
                      </Marker>
                      <Rectangle
                        bounds={
                          Array [
                            Array [
                              38,
                              -77,
                            ],
                            Array [
                              39,
                              -76,
                            ],
                          ]
                        }
                        weight={2}
                      />
                    </g>
                  </EditBounds>
                </div>
              </Map>
            </Map>
          </Connect(Map)>
        </div>
        <div
          className="ApplicationDock"
        >
          <Title>
            <div
              className="ApplicationDockTitle"
            >
              Edit region
            </div>
          </Title>
          <Dock>
            <InnerDock>
              <div
                className="InnerDock "
                style={
                  Object {
                    "height": 768,
                  }
                }
              >
                <div
                  className="block"
                >
                  <Text
                    label="Region Name*"
                    name="Region Name"
                    onChange={[Function]}
                    value="Mock Region"
                  >
                    <Group
                      id="region-name-input-0"
                      label="Region Name*"
                      name="Region Name"
                      onChange={[Function]}
                      value="Mock Region"
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          htmlFor="region-name-input-0"
                        >
                          Region Name*
                        </label>
                        <Input
                          id="region-name-input-0"
                          label="Region Name*"
                          name="Region Name"
                          onChange={[Function]}
                          type="text"
                          value="Mock Region"
                        >
                          <input
                            className="form-control"
                            id="region-name-input-0"
                            name="Region Name"
                            onChange={[Function]}
                            placeholder="Region Name*"
                            type="text"
                            value="Mock Region"
                          />
                        </Input>
                      </div>
                    </Group>
                  </Text>
                  <Text
                    label="Description"
                    name="Description"
                    onChange={[Function]}
                    value="Region description"
                  >
                    <Group
                      id="description-input-1"
                      label="Description"
                      name="Description"
                      onChange={[Function]}
                      value="Region description"
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          htmlFor="description-input-1"
                        >
                          Description
                        </label>
                        <Input
                          id="description-input-1"
                          label="Description"
                          name="Description"
                          onChange={[Function]}
                          type="text"
                          value="Region description"
                        >
                          <input
                            className="form-control"
                            id="description-input-1"
                            name="Description"
                            onChange={[Function]}
                            placeholder="Description"
                            type="text"
                            value="Region description"
                          />
                        </Input>
                      </div>
                    </Group>
                  </Text>
                  <h5>
                    Analysis bounds
                  </h5>
                  <Text
                    label="North bound"
                    name="North bound"
                    onChange={[Function]}
                    value={39}
                  >
                    <Group
                      id="north-bound-input-2"
                      label="North bound"
                      name="North bound"
                      onChange={[Function]}
                      value={39}
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          htmlFor="north-bound-input-2"
                        >
                          North bound
                        </label>
                        <Input
                          id="north-bound-input-2"
                          label="North bound"
                          name="North bound"
                          onChange={[Function]}
                          type="text"
                          value={39}
                        >
                          <input
                            className="form-control"
                            id="north-bound-input-2"
                            name="North bound"
                            onChange={[Function]}
                            placeholder="North bound"
                            type="text"
                            value={39}
                          />
                        </Input>
                      </div>
                    </Group>
                  </Text>
                  <Text
                    label="South bound"
                    name="South bound"
                    onChange={[Function]}
                    value={38}
                  >
                    <Group
                      id="south-bound-input-3"
                      label="South bound"
                      name="South bound"
                      onChange={[Function]}
                      value={38}
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          htmlFor="south-bound-input-3"
                        >
                          South bound
                        </label>
                        <Input
                          id="south-bound-input-3"
                          label="South bound"
                          name="South bound"
                          onChange={[Function]}
                          type="text"
                          value={38}
                        >
                          <input
                            className="form-control"
                            id="south-bound-input-3"
                            name="South bound"
                            onChange={[Function]}
                            placeholder="South bound"
                            type="text"
                            value={38}
                          />
                        </Input>
                      </div>
                    </Group>
                  </Text>
                  <Text
                    label="East bound"
                    name="East bound"
                    onChange={[Function]}
                    value={-76}
                  >
                    <Group
                      id="east-bound-input-4"
                      label="East bound"
                      name="East bound"
                      onChange={[Function]}
                      value={-76}
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          htmlFor="east-bound-input-4"
                        >
                          East bound
                        </label>
                        <Input
                          id="east-bound-input-4"
                          label="East bound"
                          name="East bound"
                          onChange={[Function]}
                          type="text"
                          value={-76}
                        >
                          <input
                            className="form-control"
                            id="east-bound-input-4"
                            name="East bound"
                            onChange={[Function]}
                            placeholder="East bound"
                            type="text"
                            value={-76}
                          />
                        </Input>
                      </div>
                    </Group>
                  </Text>
                  <Text
                    label="West bound"
                    name="West bound"
                    onChange={[Function]}
                    value={-77}
                  >
                    <Group
                      id="west-bound-input-5"
                      label="West bound"
                      name="West bound"
                      onChange={[Function]}
                      value={-77}
                    >
                      <div
                        className="form-group"
                      >
                        <label
                          htmlFor="west-bound-input-5"
                        >
                          West bound
                        </label>
                        <Input
                          id="west-bound-input-5"
                          label="West bound"
                          name="West bound"
                          onChange={[Function]}
                          type="text"
                          value={-77}
                        >
                          <input
                            className="form-control"
                            id="west-bound-input-5"
                            name="West bound"
                            onChange={[Function]}
                            placeholder="West bound"
                            type="text"
                            value={-77}
                          />
                        </Input>
                      </div>
                    </Group>
                  </Text>
                  <p>
                    <em>
                      OSM updates have been temporarily disabled and will be moved to the bundle management page.
                    </em>
                  </p>
                  <Button
                    block={true}
                    disabled={false}
                    name="Save changes"
                    onClick={[Function]}
                    style="success"
                  >
                    <a
                      className="btn btn-success btn-block"
                      disabled={false}
                      href="#"
                      name="Save changes"
                      onClick={[Function]}
                      tabIndex={0}
                    >
                      Save changes
                    </a>
                  </Button>
                  <Button
                    block={true}
                    onClick={[Function]}
                    style="danger"
                  >
                    <a
                      className="btn btn-danger btn-block"
                      href="#"
                      onClick={[Function]}
                      tabIndex={0}
                    >
                      <Icon
                        type="trash"
                      >
                        <i
                          className="fa fa-trash fa-fw "
                        />
                      </Icon>
                       
                      DANGER: Delete this region
                    </a>
                  </Button>
                </div>
              </div>
            </InnerDock>
          </Dock>
        </div>
      </div>
    </Application>
  </EditRegion>
</Provider>
`;
